### DISCUSSION TITLE
Your favorite Shell GPT commands

### DISCUSSION THREAD

**Discussion by TheR1D** [TheR1D/shell_gpt#53] [Show and tell]
There are many use cases for Shell GPT, such as generating shell commands, code, and more. Let's share our favorite commands in this discussion. I'll start with mine:

### 1. Say
It's really cool to use Shell GPT with the macOS default say command, which will read out the generated text for you. This is especially useful for large outputs:
```shell
say "$(sgpt 'Explain GitHub workflows')"
```
### 2. Conversions
I really like to use Shell GPT for all kinds of conversions:
```shell
sgpt "2km to miles"
# -> 2 kilometers is approximately 1.24 miles.
sgpt "2 days to seconds"
# There are 172,800 seconds in 2 days.
```
### 3. Summarization and analyzing
Since GPT-3+ models can summarize and analyze text, we can use them to explain errors, code, and more. In this first example, we're calling a Python script that does a zero division (2 / 0), redirecting stderr to stdout, and passing it to sgpt:
```shell
sgpt "explain error: $(python zero_division.py 2>&1)"
# -> This error message indicates that there was an attempt to divide a number by zero, which is not possible in mathematics. The code attempted to divide the number 2 by 0, resulting in a ZeroDivisionError. The traceback message shows where the error occurred and the line of code that caused the error.
```
In this second example, I'm passing the source code of the `sgpt` app module and asking for an explanation:
```shell
sgpt "explain following code: $(cat app.py)"
# -> This code defines a command-line interface that provides a simple interface for OpenAI's ChatGPT API using Typer. It supports different modes of output including shell commands and code, and allows users to specify the desired OpenAI model and length and other options of the output, and API Key is stored locally for easy use in future runs.
```
Here are a few other use cases:
```shell
sgpt --code "generate docstrings for my python module: $(cat app.py)"
# -> ...
sgpt --code "optimize my python module: $(cat app.py)"
# -> ...
sgpt "what is the most used word in following text: $(curl https://raw.githubusercontent.com/TheR1D/shell_gpt/main/README.md)"
# -> The most used word in the text is "sgpt"
```

**Comment by lpurdy01** [comment]
I really like using it to generate  [mermaid.js](https://mermaid.js.org/) diagrams of arbitrary code files:
Prompt:
```
$ cat sgpt/client.py | sgpt --model gpt-4 "Please summarize this python file as a mermaid class diagram"
```

Response:
```mermaid
classDiagram
    class OpenAIClient {
        -api_key: str
        -api_host: str
        -cache: Cache
        -chat_cache: ChatCache
        +__init__(api_host: str, api_key: str): None
        -_request(messages: List, model: str, temperature: float, top_probability: float): Dict
        +get_completion(messages: List[str], model: str, temperature: float, top_probability: float, caching: bool): str
    }
    class Cache {
        <<class>>
    }
    class ChatCache {
        <<class>>
    }
    OpenAIClient --> Cache
    OpenAIClient --> ChatCache
```

**Comment by lalligagger** [comment]
With some added context I have been generating satellite imagery searches with sgpt successfully. In a way I am happy that even GPT4 cannot do this reliably so I have this side project to work on. 😄 

```
(satgpt) kevinlalli@Kevins-MBP shell_gpt % sgpt -se "find me satellite images of the Honghe Hani Rice Terraces. use a 10km bounding box and search landsat 8&9 surface reflectance from summer fo 2022. save stac-client search results to ./data/rice-paddies.json"
stac-client search https://landsatlook.usgs.gov/stac-server/ --bbox 102.5 22.5 103.5 23.5 --collections landsat-c2l2-sr --datetime 2022-06-01/2022-08-31 --query "eo:cloud_cover<20" > ./data/rice-paddies.json
```

<img src="https://user-images.githubusercontent.com/11700267/229417139-89b79a4a-afb2-465d-89c9-b1274b37ed42.png" width=50% height=50%>

**Comment by xsoc** [comment]
I've wrapped most things up into shell functions or aliases in my `.zprofile` / `.bash_profile`, mostly with a focus on typing less, I'm very lazy ;)

Call `sgpt` without quotes and slightly fewer characters
```
function ai() {
    sgpt --no-cache "$*"
}
```

Call `sgpt --shell` with less characters and no quotes
```
function aido() {
    sgpt --no-cache -s "$*"
}
```

Generate code with less.... Well, you've probably figured out the pattern by now :)
```
function aicode() {
    sgpt --no-cache --code "$*"
}
```

A slightly different approach to what @TheR1D already mentioned with voice output (only Mac, sorry), this will vocalise and print to the screen.
```
function aisay() {
    sgpt --no-cache "$*" | tee >(say)
}
```

Ok so here's something a little bit different. Basically piping in the output of a directory listing and getting some guesses as to the file contents
```
alias lsai="sgpt --no-cache \"Within the \`pwd\` directory is a number of files. Please describe each of them in detail. The file names are \`ls -1aG\`\""
alias ails=lsai
```
Some sample output from the above:
```
The "." directory represents the current directory.
The ".." directory represents the parent directory.
The ".VolumeIcon.icns" file is an icon file for the volume.
The ".file" file is a hidden file.
The ".vol" file is a hidden file.
The "Applications" directory contains applications installed on the system.
The "Library" directory contains system and user libraries.
The "System" directory contains system files and resources.
The "Users" directory contains user home directories.
...
```

Lastly some interactive chat tools. ~~Personally, I only want a single, throwaway session, so I use the `temp` session name.
```
alias aichat="sgpt --repl temp"
```

*I've updated this comment to incorporate suggestions/feedback from TheR1D*

**Reply by TheR1D** [reply]
The only parameter removed from `sgpt` is `--execute`. I don't think we are making updates to parameters that might potentially break scripts, except for the mentioned `--execute`. Throwaway sessions are available in sgpt by using "temp" as the session name, as described in the README `sgpt --repl temp`.

**Reply by endolith** [reply]
I have instructions for minimal typing fish and Cmder here: https://github.com/TheR1D/shell_gpt/issues/128#issuecomment-1532285308

**Reply by moritztim** [reply]
you keep adding `--no-cache`. You could configure this as the default and then use `--cache` instead whenever you do need it.

**Comment by xsoc** [comment]
`--no-animation` got the axe recently too.
Completely missed the temp option in the docs 😅 thanks for pointing it out 😄

**Reply by TheR1D** [reply]
True, `--no-animation` as well 🙂

**Comment by will-wright-eng** [comment]
a rather verbose way to save all your transactions:

```bash
function gpt() {
    if [ -z "$2" ]; then
        echo "no session provided"
        local filename="$(date +"%Y-%m-%d")-sgpt.md"
    else
        local filename="$(date +"%Y-%m-%d")-$1-sgpt.md"
    fi
    local filedir="$HOME/sgpt"
    local filepath="$filedir/$filename"
    echo $filepath

    if [[ "$1" == "open" ]]; then
        echo "$1"
        subl "$filedir"
        return 0
    fi

    if [ -e "$filepath" ]; then
        echo "$filepath already exists"
    else
        touch "$filepath"
        echo "$filepath created"
    fi

    if [ -z "$2" ]; then
        echo '## Prompt\n\n```bash\nsgpt \"$1\"\n```\n\n## Response\n' >> "$filepath"
        sgpt "$1" | tee -a "$filepath"
    else
        echo '## Prompt\n\n```bash\nsgpt \"$2\"\n```\n\n## Response\n' >> "$filepath"
        sgpt --chat "$1" "$2" | tee -a "$filepath"
    fi
}
```

as opposed to... https://github.com/TheR1D/shell_gpt/discussions/235#discussion-5176803

*example usage*
```bash
gpt test-session "$(cat gpt-prompt-file.txt)"
```

**Comment by moritztim** [comment]
I used this in my shell theme (zsh) to generate an emoji instead of the typical `$` character at the start of the prompt
```sh
ai_emoji() {
  command sgpt --model=gpt-3.5-turbo --cache \
  "Return an emoji that could accompany this path in a shell prompt: ${PWD} and don't include any other text. If you can't think of one, instead return a symbol that would fit. Your response will be used as is in the shell prompt so provide no other output than just one symbol." \
  | sed 's/^[[:space:]]*//;s/[[:space:]]*$//'
}
```
Here's some examples:
```
╭──(/tmp/shell_gpt)───────────────────(user@moritz:pts/0)──╮
╰─()──🐚 cd ../school                            ──(16:28)─╯
╭──(/tmp/school)──────────────────────(user@moritz:pts/0)──╮
╰─()──🏫 cd ../artsncrafts                       ──(16:28)─╯
╭──(/tmp/artsncrafts)─────────────────(user@moritz:pts/0)──╮
╰─()──🎨 cd ../pictures                          ──(16:28)─╯
╭──(/tmp/pictures)────────────────────(user@moritz:pts/0)──╮
╰─()──📷                                         ──(16:28)─╯
```

**Reply by moritztim** [reply]
It's always fun to see what emoji it comes up with whenever I create a new folder. Here's the downsides of my very simple implementation:
- The caching could def be optimized, storing a simple list of paths and their emojis, so it doesn't take as long every time.
- It also sucks when the api is down and my shell kinda doesn't work anymore, I recommend adding a fallback value, should be pretty simple.
- It uses the :file_folder: emoji a lot but that's better than it saying "I'm sorry I couldn't find a fitting emoji", which has never happened with this prompt.

**Reply by moritztim** [reply]
Edit: now trimming whitespaces due to markdown formatting

**Comment by moritztim** [comment]
Generate  code and watch it live in vscodium
Works with zsh.
```sh
aicode() {
    # Check if input is available from the pipe
    if [ -t 0 ]; then
        # If no input from pipe, pass only the prompt to the AI
        sgpt --code "$@" | codium -
    else
        # If input is received from pipe, concatenate it with the provided prompt
        cat - | sgpt --code "$@" | codium -
    fi
}
```

**Comment by moritztim** [comment]
Markdown and code highlighting and copy output to clipboard.
Originally by @vansoest, modified heavily by gpt and me
Works with zsh.
https://github.com/TheR1D/shell_gpt/discussions/200#discussioncomment-7790951
https://gist.github.com/moritz-t-w/f737e41f23d529cce9fb38fa2dfc3180

**Reply by moritztim** [reply]
@TheR1D I wonder if something like this could make its way into sgpt itself...

**Comment by hcal** [comment]
I have found this meeting minutes role to be pretty useful.  
 
#.config/shell_gpt/roles/min.json
```
{
  "name": "meetingMinutesGenerator",
  "expecting": "Requests to create detailed meeting minutes in Markdown. This includes action items with owners and a summary of discussions.",
  "variables": null,
  "role": "Your task is to generate meeting minutes in Markdown format with the following structure: 1. Meeting Title at the top. 2. List of action items in bullet form, including dates and owners (list 'TBD' if unknown). 3. For each action item, list relevant chat lines or quotes from the transcript as sub-bullets. 4. Include an itemized discussion summary at the end, using Markdown subheadings for each major topic and key points, supported by direct quotes frequently. Focus on clarity, organization, and capturing the essence of the meeting."
}
```
For basic usage,  you can just cat a transcript to it like this `cat meetingTranscript.txt | sgpt --role min > meeting.md`

Using Wayland clipboard tools, I often just copy into the clipboard the file from my browser and run this `wl-paste | sgpt --role min | wl-copy -f` which puts the meeting minutes in my clip board to past into my notes app.
